cli/src/cmd/forge/script/executor.rs

```diff
@@ -0,0 +1,148 @@
+use crate::{cmd::needs_setup, utils};
+
+use ethers::{
+    solc::artifacts::CompactContractBytecode,
+    types::{transaction::eip2718::TypedTransaction, Address, U256},
+};
+use forge::{
+    executor::{builder::Backend, ExecutorBuilder},
+    trace::CallTraceDecoder,
+};
+
+use std::collections::VecDeque;
+
+use crate::cmd::forge::script::*;
+
+impl ScriptArgs {
+    /// Locally deploys and executes the contract method that will collect all broadcastable
+    /// transactions.
+    pub async fn execute(
+        &self,
+        script_config: &mut ScriptConfig,
+        contract: CompactContractBytecode,
+        sender: Option<Address>,
+        predeploy_libraries: &[ethers::types::Bytes],
+    ) -> eyre::Result<ScriptResult> {
+        let (needs_setup, abi, bytecode) = needs_setup(contract);
+
+        let mut runner = self
+            .prepare_runner(script_config, sender.unwrap_or(script_config.evm_opts.sender))
+            .await;
+        let (address, mut result) = runner.setup(
+            predeploy_libraries,
+            bytecode,
+            needs_setup,
+            self.broadcast,
+            script_config.sender_nonce,
+        )?;
+
+        let (func, calldata) = self.get_method_and_calldata(&abi)?;
+        script_config.called_function = Some(func);
+
+        let script_result = runner.script(address, calldata)?;
+
+        result.success &= script_result.success;
+        result.gas = script_result.gas;
+        result.logs.extend(script_result.logs);
+        result.traces.extend(script_result.traces);
+        result.debug = script_result.debug;
+        result.labeled_addresses.extend(script_result.labeled_addresses);
+        result.returned = script_result.returned;
+
+        match (&mut result.transactions, script_result.transactions) {
+            (Some(txs), Some(new_txs)) => {
+                txs.extend(new_txs);
+            }
+            (None, Some(new_txs)) => {
+                result.transactions = Some(new_txs);
+            }
+            _ => {}
+        }
+
+        Ok(result)
+    }
+
+    /// Executes a list of transactions locally and persists their state.
+    pub async fn execute_transactions(
+        &self,
+        transactions: VecDeque<TypedTransaction>,
+        script_config: &ScriptConfig,
+        decoder: &mut CallTraceDecoder,
+    ) -> eyre::Result<VecDeque<TypedTransaction>> {
```

Comment:
AFAICT what this function does is run the transactions and print their traces, it doesn't seem to persist state?

Reply:
it does persist the state, the function calls `.simulate(tx)` which in turn calls `Script.call(commit=true)`